File: src/main/java/me/centralhardware/znatoki/telegram/statistic/telegram/CommandHandler/CommandHandler.java
package me.centralhardware.znatoki.telegram.statistic.telegram.CommandHandler;

import me.centralhardware.znatoki.telegram.statistic.entity.postgres.Role;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.UserMapper;
import me.centralhardware.znatoki.telegram.statistic.telegram.Handler;
import me.centralhardware.znatoki.telegram.statistic.telegram.TelegramSender;
import me.centralhardware.znatoki.telegram.statistic.telegram.TelegramUtil;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.Update;

public abstract class CommandHandler implements Handler {

    @Autowired
    protected TelegramSender sender;
    @Autowired
    private TelegramUtil telegramUtil;
    @Autowired
    private UserMapper userMapper;

    public abstract void handle(Message message);

    public void handle(Update update){
        handle(update.getMessage());
    }

    public abstract boolean isAcceptable(String data);

    /**
     * @return True, if giving callbackQuery can be processed by this handler
     */
    public boolean isAcceptable(Update update){
        if (!update.hasMessage()) return false;

        String text = update.getMessage().getText();
        if (StringUtils.isBlank(text)) return false;

        return isAcceptable(update.getMessage().getText()) && checkAuth(update);
    }

    public abstract Role getRequiredRole();

    public boolean checkAuth(Update update){
        var id = telegramUtil.getUserId(update);
        var user = userMapper.getById(id);

        var requiredRole = getRequiredRole();
        if (requiredRole == null) return true;

        if (user == null || user.getRole() == Role.BLOCK){
            sendAccessDenied(update);
            return false;
        }


        switch (requiredRole){
            case ADMIN -> {
                if (user.getRole() != Role.ADMIN){
                    sendAccessDenied(update);
                    return false;
                }
            }
            case READ_WRITE -> {
                if (user.getRole() != Role.ADMIN && user.getRole() != Role.READ_WRITE){
                    sendAccessDenied(update);
                    return false;
                }
            }
        }

        return true;
    }

    private void sendAccessDenied(Update update){
        sender.sendText("Недостаточно прав", TelegramUtil.getFrom(update));
    }


}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/telegram/commandHandler/CommandHandler.kt
package me.centralhardware.znatoki.telegram.statistic.telegram.commandHandler

import me.centralhardware.znatoki.telegram.statistic.entity.Role
import me.centralhardware.znatoki.telegram.statistic.mapper.UserMapper
import me.centralhardware.znatoki.telegram.statistic.telegram.Handler
import me.centralhardware.znatoki.telegram.statistic.telegram.TelegramSender
import me.centralhardware.znatoki.telegram.statistic.userId
import org.apache.commons.lang3.StringUtils
import org.telegram.telegrambots.meta.api.objects.Update

abstract class CommandHandler(
    val sender: TelegramSender,
    val userMapper: UserMapper
) : Handler {

    abstract override fun isAcceptable(data: String): Boolean

    /**
     * @return True, if giving callbackQuery can be processed by this handler
     */
    override fun isAcceptable(update: Update): Boolean {
        if (!update.hasMessage()) return false

        val text = update.message.text
        if (StringUtils.isBlank(text)) return false

        return isAcceptable(update.message.text) && checkAuth(update)
    }

    abstract fun getRequiredRole(): Role?

    private fun checkAuth(update: Update): Boolean {
        val id = update.userId()
        val user = userMapper.getById(id)

        val requiredRole = getRequiredRole() ?: return true

        if (user == null || user.role == Role.BLOCK) {
            sendAccessDenied(update)
            return false
        }

        return when (requiredRole) {
            Role.ADMIN -> {
                if (user.role != Role.ADMIN) {
                    sendAccessDenied(update)
                    false
                } else true
            }

            Role.READ_WRITE -> {
                if (user.role != Role.ADMIN && user.role != Role.READ_WRITE) {
                    sendAccessDenied(update)
                    false
                } else true
            }

            else -> true
        }
    }

    private fun sendAccessDenied(update: Update) {
        sender.sendText("Недостаточно прав", update.userId())
    }
}